src/shared/mount-util.c

```diff
@@ -250,130 +177,154 @@ int bind_remount_recursive_with_mountinfo(
                         return log_debug_errno(r, "Failed to parse /proc/self/mountinfo: %m");
 
                 for (;;) {
+                        _cleanup_free_ char *d = NULL;
+                        const char *path, *type, *opts;
+                        unsigned long flags = 0;
                         struct libmnt_fs *fs;
-                        const char *path, *type;
 
                         r = mnt_table_next_fs(table, iter, &fs);
-                        if (r == 1)
+                        if (r == 1) /* EOF */
                                 break;
                         if (r < 0)
                                 return log_debug_errno(r, "Failed to get next entry from /proc/self/mountinfo: %m");
 
                         path = mnt_fs_get_target(fs);
+                        if (!path)
+                                continue;
+
+                        if (!path_startswith(path, prefix))
+                                continue;
+
                         type = mnt_fs_get_fstype(fs);
-                        if (!path || !type)
+                        if (!type)
+                                continue;
+
+                        /* Let's ignore autofs mounts.  If they aren't triggered yet, we want to avoid
+                         * triggering them, as we don't make any guarantees for future submounts anyway.  If
+                         * they are already triggered, then we will find another entry for this. */
+                        if (streq(type, "autofs")) {
+                                top_autofs = top_autofs || path_equal(path, prefix);
                                 continue;
+                        }
 
-                        if (!path_startswith(path, simplified))
+                        if (set_contains(done, path))
```

Comment:
Any reason this is not checked before the mnt_fs_get_fstype() call? Probably very minor, but checking our own hash table should be quicker than calling out into libmount (and possibly a syscall/fs access?) I'd imagine?

Reply:
to track `top_autofs` we need to look at the type first. hence we need to check fstype first, and only then check the list of done mounts